### DISCUSSION TITLE
Force imu 6050 to start  yaw =0 in a physical plane position after calibration process

### DISCUSSION THREAD

**Discussion by MaxESP** [jrowberg/i2cdevlib#732] [Q&A]
Hello all, I am building a tilt compensate compass with LSM303DLHC and imu 6050 .

All is good during the calibration process I have manage to store the offset before the interrupt  and my compass is working well with dmp sketch.

my problem after the calibration on a flat surface is that i want that yaw start at 0  on a physical position of the couple LSM303DLHC and imu 6050 .

I am lost is it possible to achieve this .

![compass](https://user-images.githubusercontent.com/41333143/220858626-581131f8-7c12-4ef3-9943-e79c7f286503.png)

**Comment by MaxESP** [comment]
Sorry I was wrong it's work now as expected

**Comment by MaxESP** [comment]
But the problem is that yaw drift a lot compare to roll and pitch why ?

**Comment by Aarpp** [comment]
hello did you mean you wanted yaw to always be zero on initial startup  ? 
I managed to always have yaw as 0 using a compass with tilt-comp on a UAV i made by creating a vector that stores the last input desired yaw value and storing it. happy to share IF that is what you are looking for. 
I also used sensor fusion or a simple comp filter to combine ypr[0] and tilt heading, resulting in a steady non drift yaw value. 

I apologize for not better understanding your question.

**Reply by Halvhjearne** [reply]
i would like to see example of this, if you dont mind?

**Reply by MaxESP** [reply]
hello Halvh,

You have to declare all the data variable at global variable to 0 and the couple mpu6050 and LSM303DLH_ on a flat surface to make the calibaration 
float roll =0;
float pitch=0;
float yaw=0;
float angle_yaw=0;

int deg1 = (headingDegrees1);
int  angle_yaw=yaw;
//angle_yaw -= course_deviation(angle_yaw, deg1) / 1200.0;       //Calculate the difference between the gyro and compass heading and make a small correction.
  if (angle_yaw < 0) angle_yaw += 360;                                             //If the compass heading becomes smaller then 0, 360 is added to keep it in the 0 till 360 degrees range.
  else if (angle_yaw >= 360) angle_yaw -= 360;   
 
//Serial2.printf("%i,%i\r\n",deg1,roll1);
Serial.printf("%i,%i,%i\r\n",deg1,roll1,angle_yaw);
Serial.println("");

**Comment by MaxESP** [comment]
![23 degree angle](https://github.com/jrowberg/i2cdevlib/assets/41333143/45643efe-c772-4422-b3b5-a0f0e49cacc9)
![flat](https://github.com/jrowberg/i2cdevlib/assets/41333143/650e1ece-8c9f-4e52-9440-2935701089c3)
